### PR TITLE
api-gateway: create RoleBinding attaching Role to ServiceAccount

### PR DESCRIPTION
**Pull request by nathancoleman**
**Changes proposed in this PR:**
We create a `Role` and `ServiceAccount` for the api-gateway `Deployment` in certain scenarios; however, we do not create a `RoleBinding` to bind them together.

This is causing issues when deploying to OpenShift where the `ServiceAccount` needs a `SecurityContextConstraints` resource attached, and that chain needs to go:
SecurityContextConstraints => Role => RoleBinding => ServiceAccount

**How I've tested this PR:**
Deploy into K8s with system-managed ACLs enabled and verify that the `RoleBinding` is now created alongside the `Role`, `ServiceAccount`, etc.

**How I expect reviewers to test this PR:**
See above

Checklist:
- [x] Tests added
- [ ] CHANGELOG entry added 
  > HashiCorp engineers only, community PRs should not add a changelog entry.
  > Entries should use present tense (e.g. Add support for...)

### CODE REVIEW

File: control-plane/api-gateway/gatekeeper/role.go
Comment (nathancoleman) [inline control-plane/api-gateway/gatekeeper/role.go:1]: Will consider breaking the `RoleBinding` logic out from the existing `Role` logic before review + merge

File: charts/consul/templates/connect-inject-clusterrole.yaml [hunk 1]
```diff
@@ -72,8 +72,9 @@ rules:
   - watch
   - delete
   - update
+# TODO Can we gate this so that we're only granting when necessary?
```
Comment (nathancoleman) [inline charts/consul/templates/connect-inject-clusterrole.yaml:75]: Removing out-of-scope TODO
```suggestion
```

File: control-plane/api-gateway/gatekeeper/deployment.go [hunk 2]
```diff
@@ -75,8 +75,8 @@ func (g *Gatekeeper) upsertDeployment(ctx context.Context, gateway gwv1beta1.Gat
 	return nil
 }
 
-func (g *Gatekeeper) deleteDeployment(ctx context.Context, nsname types.NamespacedName) error {
-	err := g.Client.Delete(ctx, &appsv1.Deployment{ObjectMeta: metav1.ObjectMeta{Name: nsname.Name, Namespace: nsname.Namespace}})
+func (g *Gatekeeper) deleteDeployment(ctx context.Context, gwName types.NamespacedName) error {
```
Comment (nathancoleman) [inline control-plane/api-gateway/gatekeeper/deployment.go:78]: These changes were just improving a variable name that was confusing to me coming into the project near the end

File: control-plane/api-gateway/gatekeeper/gatekeeper.go [hunk 3]
```diff
@@ -53,20 +57,27 @@ func (g *Gatekeeper) Upsert(ctx context.Context, gateway gwv1beta1.Gateway, gcc
 }
 
 // Delete removes the resources for handling routing of network traffic.
-func (g *Gatekeeper) Delete(ctx context.Context, nsname types.NamespacedName) error {
-	if err := g.deleteRole(ctx, nsname); err != nil {
+// This is done in the reverse order of Upsert due to dependencies between resources.
```
Comment (nathancoleman) [inline control-plane/api-gateway/gatekeeper/gatekeeper.go:60]: Calling this out as the reason for re-organizing the order of delete logic below

File: control-plane/api-gateway/gatekeeper/gatekeeper.go [hunk 4]
```diff
@@ -76,14 +87,14 @@ func (g *Gatekeeper) Delete(ctx context.Context, nsname types.NamespacedName) er
 // resourceMutator is passed to create or update functions to mutate Kubernetes resources.
 type resourceMutator = func() error
 
-func (g Gatekeeper) namespacedName(gateway gwv1beta1.Gateway) types.NamespacedName {
+func (g *Gatekeeper) namespacedName(gateway gwv1beta1.Gateway) types.NamespacedName {
```
Comment (nathancoleman) [inline control-plane/api-gateway/gatekeeper/gatekeeper.go:90]: We were using pointer receivers everywhere else except for this func and the one immediately below

**Review by nathancoleman — COMMENTED** [review]
Personal review

File: control-plane/api-gateway/gatekeeper/rolebinding.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	roleBinding = g.roleBinding(gateway, gcc, config)
+	if err := ctrl.SetControllerReference(&gateway, roleBinding, g.Client.Scheme()); err != nil {
+		return err
+	}
+	if err := g.Client.Create(ctx, roleBinding); err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (mikemorris) [inline control-plane/api-gateway/gatekeeper/rolebinding.go:53]: Not blocking, but this looks almost identical to the role upsert logic, would it be worth considering refactoring these into a generic upsert implementation?

**Review by mikemorris — APPROVED** [review]

File: control-plane/api-gateway/gatekeeper/rolebinding.go (hunk 5 above)
Comment (nathancoleman) [reply control-plane/api-gateway/gatekeeper/rolebinding.go:53]: It's almost identical for every resource we're creating. I'd love to refactor but don't wanna blow up the scope of this PR. We can file some work in the current epic to look into this if you're okay with that.

File: control-plane/api-gateway/gatekeeper/role.go [hunk 6]
```diff
@@ -24,27 +24,20 @@ func (g *Gatekeeper) upsertRole(ctx context.Context, gateway gwv1beta1.Gateway,
...
+	} else if !k8serrors.IsNotFound(err) {
 		// Ensure we own the Role.
 		for _, ref := range role.GetOwnerReferences() {
 			if ref.UID == gateway.GetUID() && ref.Name == gateway.GetName() {
 				// We found ourselves!
 				return nil
 			}
 		}
-		return errors.New("Role not owned by controller")
+		return errors.New("role not owned by controller")
```
[the remark this answers is not in the document]
Comment (nathancoleman) [reply control-plane/api-gateway/gatekeeper/role.go:40]: Yeah, the comment was a copy-pasta leftover from some previous work. No logic change 👍

**Pull request merged**